### DISCUSSION TITLE
Native WebGPU renderer for leaflet

### DISCUSSION THREAD

**Discussion by MatiasHiltunen** [Leaflet/Leaflet#9207] [Ideas]
Hello,

I would like to hear your opinion of this idea/proposal. 

I'm up for the challenge on implementing lightweight native WebGPU renderer for the leaflet. 

I have few years of experience working with WebGL solutions and shaders and as the WebGPU is slowly maturing, I would like to begin development with it as personal learning experience to achieve deep understanding of the native WebGPU development process. Goal is to achieve best possible performance with little-to-no library usage and keep things minimal while supporting existing features.

I'm probably going to do this "learning experience" anyways with something but the best motivation for me would be to have the work eventually on actual use and let the results benefit others aswell. 

As the WebGPU still has no support in many browsers, I could make a fallback option to WebGL. 

I'm aware that there are many forks and plugins that may already do this but I would really like to get hands-on with this and dive to the deep end so to speak.

So, let me know what you think of this and of course any tips and guidance are welcome!

Br. Matias

Slava Ukraini!

**Comment by IvanSanchez** [comment]
I've spent an awful lot of time in https://gitlab.com/IvanSanchez/leaflet.gleo , so... I'm wary of what you're attempting.

The most basic problems to solve in a WebGL/WebGPU renderer are the floating point precision issues (how are you going to deal with millimeter-precision geographical coordinates in float32 numbers?) and memory allocation issues (in Leaflet you can add any number of things to a map and then delete any one arbitrary thing).

I recommend you read through https://webgpufundamentals.org/ and ponder about the issues I pointed out.

**Reply by MatiasHiltunen** [reply]
Nice work! One of the reasons why I'm doing this is that I had to do similar workaround with canvases (listeners blocked from the underlying layer) recently while implementing Three.js layer on top of OpenLayers map to render complex 3D models on the map. To get rid off that workaround I made custom tile fetcher to load tiles directly to Three.js Plane texture which also gave a nice way to apply heightmap to the Plane aswell. 

What I'm trying to do is to create similar system with WebGPU which allows to compute pixel values from the heightmap on GPU and return the value to javaScript.

As the visualization is done mostly on GPU side that is an enormous performance boost and that enables to render hundreds of thousands, if not millions of models/points/sprites on the map to their locations if instancing is used (positions set to instance matrix)

I'm not worried about float32 values as thats more of an scaling issue with level of detail and visualization is always more or less aproximation. Actual correct values are always kept in js memory.

**Reply by IvanSanchez** [reply]
So what's your goal with this, what's the use case? I assumed you were planning to create a replacement for `L.Renderer`, for lines/polygons/circles.

If the use case is calculate height maps and so, I've already got a whole architecture to deal with scalar fields and do pixel queries with WebGL1. If the use case is 3D models, remember that Leaflet is strictly 2D.

**Reply by MatiasHiltunen** [reply]
Polygons, Lines, Points etc. would surely benefit from this a great deal as it can be currently seen on any webgl-based map. 
I can imagine a lot of usecases other than 3D, performance and enhanced user experience being one. Would be nice to throw 1 000 000 points on the map while everything still running smoothly at 60fps, without blocking the main thread. 

I have enjoyed working with leaflet library very much for years but there have been a lot of cases where I have had to choose commercial gl-based map instead solely because of performance.

I have to check out your solutions, getting pixel values from GPU memory with Webgl 1.1 is not a trivial task! Would like to hear more from that! 

WebGPU offers a way to return computed values from shaders back to js without hacks that webgl 1 and 2 required.

With such parallel computing power extracting the data from data-tiles (geotiff, terrarium etc.) would make a massive difference compared to js-calculated values. I'm not saying that this should be a feature here, just that it would open a lot of possibilities.

With your existing solutions you're probably able to change color values of the tiles with shader code? It is a nice way to have different color styles on raster basemap tiles, for example OSM tiles can be changed to basically any color while maintaining the quality and performance.

**Reply by MatiasHiltunen** [reply]
If raster tile basemaps are recolored efficiently on client side that would be very sustainable solution atleast as that would reduce load on tileservers. One example here with OSM raster tiles with WebGL tiles using OpenLayers and custom color adjustments for testing purposes![IMG_20240106_005515.png](https://github.com/Leaflet/Leaflet/assets/26164588/b107fd14-697a-4b8a-b3af-d5b511d2402c)

**Reply by IvanSanchez** [reply]
> getting pixel values from GPU memory with Webgl 1.1 is not a trivial task

It kinda is, once you get framebuffer references and byte alignment right. https://gitlab.com/IvanSanchez/glii/-/blob/master/src/FrameBuffer/FrameBuffer.mjs?ref_type=heads#L346

> With your existing solutions you're probably able to change color values of the tiles with shader code?

Been there, done that. https://gitlab.com/IvanSanchez/Leaflet.TileLayer.GL/

**Reply by MatiasHiltunen** [reply]
Well, I wouldn't call that trivial ;) Nice work!

**Reply by MatiasHiltunen** [reply]
Of course there is no point to invent wheel again and WebGPU has a long way to replace webgl. Nonetheless, I'm probably going to do this anyway as the WebGPU is something that I'm going to focus on in the future.

WebGL solution for reading/calculating pixel values from tiles works nicely but what I would like to get is the ability to use raycaster to get a value from single pixel on texture and that is where WebGPU would be usefull, that probably would not benefit leaflet as 2D map rarely needs that ability.

Appreciate the good work you have done with webgl here!

**Reply by Falke-Design** [reply]
I think WebGL / WebGPU is a important topic if we want to keep Leaflet as a solution in the web. Also I think it will be easier to implement after the refactoring of `v2`.

> I'm probably going to do this anyway as the WebGPU is something that I'm going to focus on in the future.

Yes please. If we will not add it to Leaflet core, you can provide it as a plugin anyway.

@IvanSanchez are there reasons why we would not want to add GL to Leaflet?

**Reply by IvanSanchez** [reply]
> @IvanSanchez are there reasons why we would not want to add GL to Leaflet?

Code complexity. The API differences between SVG/Canvas and GL are **really** big. With GL there's a need to *either* dynamically allocate GPU RAM (the Gleo approach) or go full vector tiles (the MapboxGL/MapLibre approach), and either mean _lots_ of code.

I should give another shot at making Leaflet-Gleo more comprehensive: automagically convert `L.Marker`s into gleo `Sprite`s and `L.Polyline`s into gleo `Stroke`s and `L.TileLayer`s into gleo `MercatorTiles` and so on. It should help with some use cases, but at the same time there would be lots of plugin incompatibilities.

I mean, I did try to [bring GL to Leaflet](https://gitlab.com/IvanSanchez/Leaflet.gl)... OMG, 8 years ago. And I hit the floating point precision problems, hard. The outcome of all that is Gleo.

**Reply by MatiasHiltunen** [reply]
@IvanSanchez WebGPU has currently abstract-numerics available, which are 64bit floats or 64bit signed integers! https://google.github.io/tour-of-wgsl/types/abstract-numerics/ 

Drawback might be that those are calculated on cpu at compile time.

On WebGPU github there is open issue to add support for 64bit floats

**Comment by MatiasHiltunen** [comment]
I have made some progress. I started with a WebGPU tile layer plugin, in which I now have ability to load tile images as textures to create a basemap. I used Gridlayer as a starting point to enable support for existing features as much as possible and of course to have already existing features available.

Few things to note:

To enable usage of cross-origin images as textures, I used WGSL 'loadExternalTexture' feature which has limited ability to access its data (have not really looked much into that yet). 

Another thing is that 'loadExternalTexture' takes currently, to my knowledge, only VideoFrame or video as uniform value. That might sound odd but that actually gives more flexibility because with VideoFrame it is possible to use zero-copy approach and videoframe also has convenient "close"-method which cleans up that from memory.

**Reply by MatiasHiltunen** [reply]
I did had to override current way of the screenspace panning to have it entirely implemented in WebGPU side.

**Reply by MatiasHiltunen** [reply]
Currently zoom and panning are working with WebGPU tile layer, next I'm trying to get MSAA to work so that quality of raster tiles can be improved. 

After that I'll publish first demo, probably in couple of weeks

**Reply by MatiasHiltunen** [reply]
Just some notes of the current status.

Initially I was going to have render calls for each tile but creating a texture atlas or implementing batch rendering would greatly reduce number of render calls and would be a much more efficient way to render basemap in this case. Goal is still to be able to use zero-copy approach. Any suggestions how this would be optimal to create are welcome!

Using fetch to get image data as buffer straight to VideoFrame should allow very efficient way to handle tile images if creating a VideoFrame from buffer does not make a copy of the data loaded from network. Fetch API allows probably few benefits over Image element as the Response object from fetch handles the data in separate thread (like with .json method). VideoFrame as extranal texture is zero-copy by design and should allow very efficient way to handle tile images. 

Results so far are engourging and I'm going to continue working on this and goal is eventually to have all the basic features implemented with WebGPU.

**Reply by Falke-Design** [reply]
Sadly I have not the experience to help you with that but I would be very interested to read through your code.

**Comment by almamigratoria-netizen** [comment]
Unless you're betting on better browser-market penetration in the future, it's tough to see (aside from the learning experience) why WebGPU tiles might be desireable.   I mean, __I'm__ very interested in seeing what you come up with.  I'm currently stuggling with mapping x,y to lat/lng for doing GOES full disk image overlays with WebGL.  The distortion wasn't so bad, but panning and zooming the map without a lot of flicker is kicking my ass, so I'd love to see a good implementation so I could do, for instance, a WebGPU version of [leaflet-velocity](https://github.com/onaci/leaflet-velocity)

**Reply by MatiasHiltunen** [reply]
I completely forgot about this! I did manage to make the rendering work somewhat nicely but had to pause the development because I had to shift my focus elsewhere. Setup for the rendering pipeline required quite extensive changes and with some parts of the leaflet basically a full rewrite. I think I can finish the prototype with minor tweaks. Now that WebGPU has greatly stabilized and new browser APIs have emerged since, I think I could make the rendering to work much better.

**Reply by yohanboniface** [reply]
Hi @MatiasHiltunen did you finally pushed some code out of your experiments ? :)

**Comment by almamigratoria-netizen** [comment]
I'd be greatly interested in seeing where you're at with this.  You should
go ahead and post to github... i'd be interested in collaborating with you
to make this leaflet 2.0 compatible.  I've got a pull to leaflet for three
leaflet version 2 plugins, and am working on a webgl windy layer, but going
WebGPU (with WebGL fallback) would be fun, I think.

**Comment by MatiasHiltunen** [comment]
Very barebones demo here, decided to leave out the leaflet completely for now as it would have not made sense to use DOM at all with the canvas based webgpu rendered map, this was the only way I find this to be useful.. https://github.com/MatiasHiltunen/webgpu_map, live demo: https://matiashiltunen.github.io/webgpu_map/

**Reply by MatiasHiltunen** [reply]
@yohanboniface @almamigratoria-netizen

**Comment by MatiasHiltunen** [comment]
Added some more features to demo (with AI) to give a some taste what the webgpu/gl can do for the rendering ![Screenshot_20260426_130532_Chrome.jpg](https://github.com/user-attachments/assets/093c53d6-0714-41b7-8205-aa6857ac4d83)

**Reply by MatiasHiltunen** [reply]
Basemap shader processing in realtime offers quite powerfull way to implement custom theming for example with the standard OSM basemap

**Reply by IvanSanchez** [reply]
Pal, I did the same at https://gitlab.com/IvanSanchez/Leaflet.TileLayer.GL/ 9 years ago. There's just no demand for that kind of feature.

**Reply by MatiasHiltunen** [reply]
@IvanSanchez Yes your work is exceptionally good and it is a shame really that the feature has not gained more traction! I happen to know that there is, in fact, need for features like this even with my own work. Btw your demo shows me a blank page.

**Reply by MatiasHiltunen** [reply]
Also the overall implementation is fundamentally different to yours @IvanSanchez

**Reply by MatiasHiltunen** [reply]
Other than theming and toying around, this could be used for in case, let's say you are looking for something on a map based on a certain color range, selecting the target color with some threshold and using post processing to add bloom effect on that said feature allows you to find quickly on the map the areas with similar color values. Didn't have a satellite basemap rightaway but this should demonstrate what I'm trying to achieve

<img width="1920" height="1032" alt="nop" src="https://github.com/user-attachments/assets/10504f84-6ab7-4fd7-82b0-c751a0248f45" />
<img width="1920" height="1032" alt="op" src="https://github.com/user-attachments/assets/380a70f0-d48d-4ddf-b7e5-f602e3ea1655" />
